database.go

```diff
@@ -324,9 +326,11 @@ func (d *Database) QueryRowContext(ctx context.Context, query string, args ...in
 }
 
 func (d *Database) queryFactory() exec.QueryFactory {
-	if d.qf == nil {
-		d.qf = exec.NewQueryFactory(d)
-	}
+	d.qfOnce.Do(func() {
+		if d.qf == nil {
```

Comment:
Nice addition! I think we could get rid of the `d.qf == nil` since the `Do` method will only execute one time.  The same logic should be applied to the `TxDatabase`.